src/Generic/Data/Internal/Enum.hs

```diff
@@ -300,3 +348,52 @@ instance (GBounded f, GBounded g) => GBounded (f :+: g) where
 instance (GBounded f, GBounded g) => GBounded (f :*: g) where
   gMinBound = gMinBound :*: gMinBound
   gMaxBound = gMaxBound :*: gMaxBound
+
+-- | Generic representation of 'Ix' types.
+--
+class GIx f where
+  gRange :: (f p, f p) -> [f p]
+  gUnsafeIndex :: (f p, f p) -> f p -> Int
+  gInRange :: (f p, f p) -> f p -> Bool
+
+instance GIx f => GIx (M1 i c f) where
+  gRange (M1 m, M1 n) = map M1 $ gRange (m, n)
+  gUnsafeIndex (M1 m, M1 n) (M1 i) = gUnsafeIndex (m, n) i
+  gInRange (M1 m, M1 n) (M1 i) = gInRange (m, n) i
+
+instance (GEnum StandardEnum f, GEnum StandardEnum g) => GIx (f :+: g) where
```

Comment:
This seems to not actually depend on the fact that the type is a sum `f :+: g`. Instead, this means that `Ix` can be derived for any `Enum` using the code in this instance. So I think this instance should really be three toplevel functions specifically for deriving `Ix` via `Enum`. In particular, that doesn't involve generics anymore, so the `g` prefix doesn't make sense.  `enumRange`, `enumIndex`, `enumInRange`, `enumRangeSize`.

A better instance for `f :+: g`, if it makes sense at all, would probably also require `f` and `g` to be both `GIx` and `Bounded`.

Reply:
First I do not think adding a `Bounded` constraint is necessary because the bounds are passed as an argument to `range`, `index` and `inRange` respectively.

As I understand you, you have the following top level definitions in mind:
```
enumRange :: Enum a => (a, a) -> [a]
enumIndex :: Enum a => (a, a) -> a -> Int
enumInRange :: Enum a => (a, a) -> a -> Bool
```
where e.g. `enumInRange` is defined as
```
enumInRange (m, n) i = i_m <= i_i && i_i <= i_n
  where
    i_m = fromEnum m
    i_n = fromEnum n
    i_i = fromEnum  i
```
I thought about that, but it felt wrong to me because of the following circumstance: The following derivation of instances for a tuple is perfectly reasonable where `Ix` is derived as defined in the standard (Haskell Report) and `Enum` is derived via `FiniteEnumeration`
```
import Data.Word
data Tuple = MkT Word8 Word8
  deriving (Eq, Ord, Ix, Generic)
  deriving Enum via (FiniteEnumeration Tuple)
```
However then we have that 
```
    inRange     (MkT 1 2, MkT 2 3) (MkT 2 1) == False
 /= enumInRange (MkT 1 2, MkT 2 3) (MkT 2 1) == True
```
The point is that `Ix` as defined in the Standard for a single constructor can skip elements in the enumeration and still index the remaining elements continuously, i.e. `range (m, n)` can be a proper subsequence of `[m .. n]`. In particular the skipped elements depend on the bounds. But `enumRange`, `enumIndex` and `enumInRange` do not skip any elements.

Thus defining `Ix` via `Enum` only really works for the case that `Enum` was derived as defined in the standard, i.e. via a `StandardEnum`. In this case the standard defines that no elements are skipped by the generically derived `Ix` instance, as it has to agree with the generically derived `Enum` instance.

I guess we could nonetheless add `enumRange`, `enumIndex` and `enumInRange` if we add a warning that describes how the generic `Ix` instance for enumerations differs from the generic `Ix` instance for *single-constructor datatypes* in the standard. Otherwise I find it a bit too misleading wrt nonstandard Enum instances.

I wanted to implement the deriving specification for `Ix` from the report such that the same newtype `Generically` can be used to derive instances for both the case of an enumeration type with only nullary constructors and the case of a single constructor type.
Wouldn't this be the expectation of someone that is used to deriving `Ix`?

As explained above the two cases are indeed so fundamentally different that they may warrant their own respective top level functions. So if `enumRange`, `enumIndex` and `enumInRange` do not skip elements in the enumeration and presume an `Enum` instance, then should `grange`, `gindex` and `ginRange` only work for single constructor types? Then there have to be two different newtypes for `DerivingVia`. Or should they work for both *only nullary constructors types* and *single constructor types*? Then how would we implement this? We cannot reuse `enumRange` etc. because requiring an `Enum` instance fails for some *single  constructor types*.

I agree that the `g` seems unnecessary.
However an alternative definition like the following does not play well with instance resolution.
```
instance GEnum StandardEnum f => GIx f where
```
Matching on `:+:` avoids that and delegates the case of a single constructor with a single field to the instance for `U1` that is not only used for enumerations but also for the single constructor scheme. While this trick solves the problems with instance resolution, it is unfortunately confusing to read this code.

I failed to implement the following instance, even under the assumption that there are only nullary constructors
```
instance (GIx f, GIx g) => GIx (f :+: g) where
```
An initial idea I had was to detect the case of only nullary constructors with a type parameter
```
data IxOnlyNullaryConstructors
data Ix
instance GIx p f => GIx p (M1 i c f) where
instance (GIx p f, GIx p g) => GIx IxOnlyNullaryConstructors (f :+: g) where
instance (GIx Ix f, GIx Ix g) => GIx Ix (f :*: g) where
instance GIx p U1 where
instance Ix c => GIx Ix (K1 i c) where
```

Sure, there may be an implementation for 
```
instance (GIx f, GIx g, GBounded f, GBounded g) => GIx (f :+: g) where
```
Is that what you meant? If you do not want to separate the case of only nullary constructors via a type parameter, then I currently do not think this is better. I fear mixing the two cases of *only nullary constructors* and *a single constructor* from the language report will be confusing and unintuitive.

On the other hand `GEnum StandardEnum` coincides with the language report for the case of deriving `Ix`. If we track the cases with a type parameter we can replace the `GEnum StandardEnum` constraint for `f` and `g` respectively in the `GIx` instance for `:+:` by the constraints `GIx` and `GBounded` in the case of an enumeration.

Sorry for the wall of text :(

Reply:
Thanks for the explanation, I think it makes sense. If I understand correctly, derivation of `Ix` (both the standard and your generic version here) is restricted to sums of nullary constructors, and single constructors (i.e., products). There doesn't seem to be a good way to do things any differently (as you point out, enabling deriving via `Enum` seems risky), so it's okay to just have one simple `GIx` class.

Reply:
I only see a risk for confusion if there is a custom `Enum` instance for product types (e.g. derived via `FiniteEnumeration`), which seems rather unlikely. This risk can be reduced by documentation. Deriving `Ix` via the `Enum` instance is well behaved in the sense that it satisfies the `Ix` laws. The reason why I did not implement a toplevel `enumRange` etc. is because 
1. I personally have no use case for it
2. There is no code reuse. Reusing `enumRange`, `enumIndex` and `enumInRange` for this instance of `:+:` seems questionable/unfeasible. It should be possible to derive `Ix` via Generically without having an `Enum` instance.
3. the risk of confusion is especially high in the context of this library, as it also exports `FiniteEnumeration`
4. it does not use GHC.Generics (If you don't think that matters, then ignore this one. I only think it mildly matters for the discoverability of this feature. I don't think people would expect this feature in a library for generics)

in that order of importance.

Reasons 2 and 4 can be solved by changing the signatures to something like
```
enumRange :: (Generic a, GEnum StandardEnum (Rep a)) => (a, a) -> [a]
enumIndex :: (Generic a, GEnum StandardEnum (Rep a)) => (a, a) -> a -> Int
enumInRange :: (Generic a, GEnum StandardEnum (Rep a)) => (a, a) -> a -> Bool

finiteEnumRange :: (Generic a, GEnum FiniteEnum (Rep a)) => (a, a) -> [a]
finiteEnumIndex :: (Generic a, GEnum FiniteEnum (Rep a)) => (a, a) -> a -> Int
finiteEnumInRange :: (Generic a, GEnum FiniteEnum (Rep a)) => (a, a) -> a -> Bool
```
Reasons 1 and 3 remain. But none of the reasons are strong arguments against ever implementing `enumRange`, `enumIndex` and `enumInRange`.

Notably deriving `Ix` via `Enum` is implemented [elsewhere](http://hackage.haskell.org/package/utility-ht-0.0.14/docs/Data-Ix-Enum.html) already.

Reply:
Yes, I agree entirely!